### PR TITLE
VReplication: Handle reference tables in sharded->sharded MoveTables

### PR DESCRIPTION
**Pull request by mattlord**
## Description

When using [`MoveTables`](https://vitess.io/docs/reference/vreplication/movetables/) (or [`Reshard`](https://vitess.io/docs/reference/vreplication/reshard/)) where the source keyspace is _sharded_ AND _you're moving 1 or more [reference tables](https://vitess.io/docs/user-guides/vschema-guide/advanced-vschema/#reference-tables)_ to the target keyspace, we would fail to create the `_reverse` workflow on the [`SwitchTraffic`](https://vitess.io/docs/reference/vreplication/movetables/#switchtraffic) step (and causing `vtctld` to panic/crash), blocking progress for the workflow.

Please [see the issue for additional details](https://github.com/vitessio/vitess/issues/13160), including a manual test case which fails before this PR but works on this PR branch.

>**Note**: after fixing this I realized that VDiff did not work as expected with Reference tables either IF there were a different number of shards between the source and target. This is because the side with more shards would have extra rows as each shard has the table in full. I added some special handling for the specific scenario so that as long as there are no mismatched rows and each shard has the same number of rows, then we clear any reported differences.

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/13160

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by mattlord** [comment]
All existing tests pass with these changes. I will now update the vreplication e2e tests and add a reference table.

**Review by rohit-nayak-ps — APPROVED** [review]

**Review by GuptaManan100 — APPROVED** [review]
LGTM!

File: go/vt/vttablet/tabletmanager/vdiff/workflow_differ.go [hunk 1]
```diff
@@ -63,7 +64,38 @@ func newWorkflowDiffer(ct *controller, opts *tabletmanagerdatapb.VDiffOptions) (
 // by MySQL on each side then we'll have the same number of extras on
 // both sides. If that's the case, then let's see if the extra rows on
 // both sides are actually different.
-func (wd *workflowDiffer) reconcileExtraRows(dr *DiffReport, maxExtraRowsToCompare int64) {
+func (wd *workflowDiffer) reconcileExtraRows(dr *DiffReport, maxExtraRowsToCompare int64) error {
+	if dr.MismatchedRows == 0 {
+		// Get the VSchema on the target and source keyspaces. We can then use this
+		// for handling additional edge cases, such as adjusting results for reference
+		// tables when the shard count is different between the source and target as
+		// then there will be a extra rows reported on the side with more shards.
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/vdiff/workflow_differ.go:72]: nit: "then there will be extra rows reported" (Omit the indefinite article)

File: go/vt/vttablet/tabletmanager/vdiff/workflow_differ.go [hunk 2]
```diff
@@ -63,7 +64,38 @@ func newWorkflowDiffer(ct *controller, opts *tabletmanagerdatapb.VDiffOptions) (
...
+		}
+		tgtvschema, err := wd.ct.ts.GetVSchema(wd.ct.vde.ctx, wd.ct.vde.thisTablet.Keyspace)
+		if err != nil {
+			return err
+		}
+		svt, sok := srcvschema.Tables[dr.TableName]
+		tvt, tok := tgtvschema.Tables[dr.TableName]
+		if dr.ExtraRowsSource > 0 && sok && svt.Type == vindexes.TypeReference && dr.ExtraRowsSource%dr.MatchingRows == 0 {
+			// We have a reference table with no mismatched rows and the number of
+			// extra rows on the source is a multiple of the matching rows. This
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/vdiff/workflow_differ.go:85]: I was wondering about the `%`. This is a nice comment that explains it.

File: go/vt/wrangler/traffic_switcher.go [hunk 3]
```diff
@@ -1309,11 +1309,25 @@ func (ts *trafficSwitcher) createReverseVReplication(ctx context.Context) error
...
+					case vindexes.TypeReference:
+						// For reference tables there are no vindexes and thus no filter to apply.
+					default:
+						// For non-reference tables we return an error if there's no primary
+						// vindex as it's not clear what to do.
+						if len(vtable.ColumnVindexes) > 0 && len(vtable.ColumnVindexes[0].Columns) > 0 {
+							inKeyrange = fmt.Sprintf(" where in_keyrange(%s, '%s.%s', '%s')", sqlparser.String(vtable.ColumnVindexes[0].Columns[0]),
+								ts.SourceKeyspaceName(), vtable.ColumnVindexes[0].Name, key.KeyRangeString(source.GetShard().KeyRange))
+						} else {
+							return vterrors.Errorf(vtrpcpb.Code_INTERNAL, "no primary vindex found for the %s table in the %s keyspace",
```
Comment (deepthi) [inline go/vt/wrangler/traffic_switcher.go:1327]: nit:
My preference is for errors to be phrased as "no primary vindex found for table %s in keyspace %s". Perhaps the current phrasing is consistent across occurrences. In which case, consistency wins.

File: go/vt/wrangler/vdiff.go [hunk 4]
```diff
@@ -262,6 +264,27 @@ func (wr *Wrangler) VDiff(ctx context.Context, targetKeyspace, workflowName, sou
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
+	// Get the VSchema on the target and source keyspaces. We can
+	// then use this for handling edge cases, such as adjusting
+	// results for reference tables when the shard count is
+	// different between the source and target as then there will
+	// be a extra rows reported on the side with more shards.
```
Comment (deepthi) [inline go/vt/wrangler/vdiff.go:271]: same nit as before: omit `a`.

**Pull request merged**